File: 1696 - Jump Game VI/solution.cpp
class Solution {
public:
    int maxResult(vector<int>& A, int k) {
        deque<int> q{0};
        for(int i = 1; i < A.size(); i++){
            if(q.front() < i-k) q.pop_front();
            A[i] += A[q[0]];
            while(!q.empty() && A[i]>A[q.back()]) q.pop_back();
            q.push_back(i);
        }
        return A.back();
    }
};

File: 1696 - Jump Game VI/solution.py
from collections import deque
class Solution:
    def maxResult(self, A: List[int], k: int) -> int:
        q = deque([0])
        for i in range(1, len(A)):
            if q[0] < i-k:
                q.popleft()
            A[i] += A[q[0]]
            while q and A[i] > A[q[-1]]:
                q.pop()
            q.append(i)
        return A[-1]